### ISSUE TITLE
Is it possible to share loaded packages accross Rserve sessions ?

### ISSUE THREAD

**Issue by fviale** [s-u/Rserve#42]
I'm new to RServe which I use on linux via a Java API. It's RServe 1.8.1

I noticed that across different sessions, needed R packages are each time reloaded (which takes time).

I wonder if there are various possibilities to load packages once for all and share them across all sessions.

Some guesses : 
1) As I start RServe from R, I could maybe load libaries before calling run.Rserve, but that would not be a good solution as in my application, I don't know before starting the server which packages will be needed
2) I noticed the function self.ctrlEval, could it be called from a session to load on the server some packages which will be then accessible to subsequent sessions ?

**Comment by s-u** [comment]
Yes, you can use `eval` and/or `source` directives in the configuration file (or corresponding start-up arguments) to run any code before the server starts up. Any packages you load that way will be pre-loaded for all subsequent connections.

Note that some packages/libraries cannot be forked (e.g., Java or anything that keeps an open connection) so you cannot pre-load those.

`self.ctrlEval` is another alternative, but control commands are broken in the 1.8 series and cannot be used i OCAP mode, so I would recommend using the pre-loading at startup.

**Comment by fviale** [comment]
Hello Simon, thanks a lot for your quick answer.

I will investigate the preloading option you mention. Actually, I was not correct when I said I used RServe 1.8.1.

I actually use RServe downloaded from CRAN (1.7-3), but connect to it in Java using Rserve-1.8.1.jar

Do you think it can cause issues to have separate versions for the server implementation and the Java API ? 
I would like to keep this version of the API, but I can build RServe 1.8.x from sources if you advise me to.

**Comment by fviale** [comment]
I managed to preload packages as I needed by executing library() calls before calling the run.Rserve function, but I have a second problem implying data-sharing. In my application performance is mandatory and I need to ensure that all computations run within a connection don't loose any loading time.

I am using a cache environment stored inside a preloaded package, and try to modify within a Rserve connection (I make sure no other concurrent connection can be run at the same time as this update). The cache update operation runs without error, but subsequent reading operations on this modified environment lead to the following error:
"lazy-load database 'P' is corrupt"

I have the feeling this is due to the way Rserve spawns connections using "copy-on-write" fork primitive.

So my second question is: is there a way to preload data for a whole RServe instance and be able to update this data within a connection (making sure no other simultaneous connexion can run at the same time) ?

I saw in the RServe changelog a mention of a "fork here" configuration which seems to open the door to updatable shared memory. Is it the case or do I have no other choice than restarting the Rserve server at each data update ?

**Comment by s-u** [comment]
No. Since any subsequent connection is already separate form the server process, you have no way to influence the server process anymore since you're divorced from it. What you see is that the server process has already opened the file so if you modify it, you're corrupting it since the server has already read portions of the file so later modifications will only break it. If you didn't open the database in the server process then it would work. If it's just lazy-load then it's sort of pointless, since you could have simply loaded it directly anyway (all that lazy-load does is to keep an offset into a file that contains serialized R objects, so you're really better of load loading R objects directly).

The only way to influence the server session are control eval/source commands. Those are synchronized, i.e. you don't need to worry about that. However, they only affect subsequent connections (by design since you're simply changing the workspace of the server process that clients are forked form). If you intend to use control commands, make sure you use the Rserve 1.7-x version from CRAN and not the more recent versions from RForge.net.

**Comment by fviale** [comment]
Following what you said, I am trying to experiment with control commands but it seems kind of locked. I'm using Rserve 1.7-3 CRAN version
Here is how I start the RServe server from the R session:
```R
library(Rserve);Rserve(debug=FALSE,port=6412,args = "--RS-conf '/tmp/Rserv.conf' --RS-enable-control  --no-save")
```
Here is the content of /tmp/Rserv.conf:
```
control enable
r-control enable
port 6412
encoding utf8
```
But then when I'm inside the RServe R session, impossible to call the self commands:
```R
Rserv started in daemon mode.
> self.ctrlEval("a <- 1")
Erreur dans self.ctrlEval("a <- 1") : 
  R control is not premitted in this instance of Rserve
> self.ctrlSource('/opt/proactive/default/addons/preload.R')
Erreur dans self.ctrlSource("/opt/proactive/default/addons/preload.R") : 
  R control is not premitted in this instance of Rserve

```
If I call the self commands via a java RConnection, the RServe log shows the same message "R control is not premitted in this instance of Rserve"

I had a look at the code of branch 1.7 but couldn't find the reason why the "self_control" flag is not set properly by my configuration file.

Do you have any idea about this issue ?

**Comment by s-u** [comment]
From Java you want to use `serverEval()` or `serverSource()` methods - those issue the corresponding control commands.

FWIW you only need the config file so you can simply use

```
R CMD Rserve --RS-conf /tmp/Rserv.conf
```

or in R

```r
Rserve::Rserve(args = "--RS-conf /tmp/Rserv.conf")
```

Note that the R session you start Rserve from is completely irrelevant - `Rserve()` is just a convenience function to start Rserve from R but it's unrelated to the session (it just calls `system()`). If you want to run Rserve in the R session that would be `run.Rserve()` instead (but that's more for testing than anything else).

**Comment by fviale** [comment]
Maybe it's better if I explain a bit more what I'm trying to achieve, and the background. Basically, what I'm trying to do is a JSR 223 Scripting Engine with a RServe backend. We integrate this script engine in our distributed workflow scheduler to allow the execution of R tasks, among other script engines available. 

Previously, we developped a R JSR223 Script engine based on JRI. It was sufficient in most scenarios, but we faced an issue regarding multiple instances of JRI engines running on the same machine and each loading large amount of data. The big data was duplicated as many time as there were concurent JRI engines.

RServe can theoretically solve this issue, as the big data can be stored and updated on the server side, while user sessions simply consume this data without replicating it. That's why we started to work on a RServe integration.

I made a diagram describing the general use case and the two options I have to send control commands

![rserve ctrl](https://cloud.githubusercontent.com/assets/3063713/10666573/b63a3ec6-78d2-11e5-9194-6a21c813bbb7.png)

So basically, what you are saying is that the self.CtrlEval() possibility will not work and I have to rely on the serverEval() java call ?

Regarding the call to Rserve(), I was using it for testing indeed, to be able to view what's happening on the server. When I run the Rserve() command, RStudio seems able to attach automagically to the spawned RServe session (running in a different process). And when I submit expressions from my JSR223 engine, I see them appearing on the RStudio-RServe session (practical for debugging). Now of course, I don't know precisely what's happening in the backend.
